core/string/print_string.h

```diff
@@ -59,7 +59,14 @@ void remove_print_handler(const PrintHandlerList *p_handler);
 extern void __print_line(String p_string);
 extern void __print_line_rich(String p_string);
 extern void print_error(String p_string);
-extern void print_verbose(String p_string);
+extern bool is_print_verbose_enabled();
+
+// This version avoids processing the text to be printed until it actually has to be printed, saving some CPU usage.
+#define print_verbose(m_text)           \
+	{                                   \
+		if (is_print_verbose_enabled()) \
+			print_line(m_text);         \
+	}
```

Comment:
That will still execute the string when called. The idea is that what creates the string is not executed until the error is printed, similar to the ERR_* macros.

Reply:
v 2.0:

```C++
// Concerned callers can use this with plain old `char *` strings instead of `print_verbose()` with `String`s to avoid the format being processed if verbose printing is disabled.
template <typename... VarArgs>
void printf_verbose(const char *p_fmt, const VarArgs... p_args) { // Notice the `f`.
    if (is_print_verbose_enabled()) {
        print_line(vformat(p_fmt, p_args...));
    }
}
```

Reply:
Its the same thing, it can't be done with a template. This is one of the good thing about macros. It copies things verbatim instead of following the code logic.

Reply:
This version won't construct any `String` if you pass plain old C string literals to it, unless verbose print is enabled.

Reply:
In any case, I can verify further and make a follow-up PR. This shouldn't be a hurdle for approval.

Reply:
Let me explain better, the problem is not so much that, but the fact that with a macro, *no code will be executed at all*, even if its complex code, as long as verbosity is disabled:

Example if you have something like this:

```
print_verbose("Some value"+ itos(someclass.get_some_value()));
```

When verbose mode is _disabled_, if `print_verbose` is a template or a function, it will first execute `"Some value"+ itos(someclass.get_some_value())` and call the function, meaning this code will run regardless of whether verbosity is enabled or not.

If `print_verbose` is a macro, then it will flatten to this:
```
// Some code
if (is_print_verbose_enabled()) {
        print_line("Some value"+ itos(someclass.get_some_value()));
}
```
This means, that `"Some value"+ itos(someclass.get_some_value())` does not get execute if verbosity is disabled. This is why a macro is preferred than a template.

Reply:
Thanks for taking the time to explain it.

However, please note that my solution is a _formatted_ print approach, you'd call this way:
`verbose_printf("Some values: %s %d", obj.get_some_str_value(), obj.get_some_int_value());`

And that's the other advantage, that it also allows to get rid of the string concat way to building messages.

Reply:
@RandomShaper I see your point, and I understand it may get rid of the string concat, but If you call a function (and this includes passing a String, Stringname, etc. that has a constructor.. I think the C++ compiler is _still_ obligued to call it before entering the `if ()` statement.

Reply:
That's the limitation of my method, as stated in the comment in the examples: the only string you should use with it is plain `char *`, including literals. Probably there's some modern C++ way of avoiding copies entirely (perfect forwarding maybe?) regardless the types.

Let's not loose sleep on this. I'll experiment a bit to try to find the silver bullet. If I can find one, I'll PR it.